### PR TITLE
schedule:  random pick hot region from store

### PR DESCRIPTION
**Pull request by nolouch**
extract hot cache and add RandomHotRegionFromStore interface.

### CODE REVIEW

**Comment by disksing** [comment]
Does this PR contain any feature change or just code refactor?

File: server/schedule/hot_net.go [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+type FlowKind uint32
+
+// Flags for flow.
+const (
+	WriteFlow FlowKind = iota
+	ReadFlow
+)
+
+// HotSpotNet is a cache hold hot regions.
+type HotSpotNet struct {
```
Comment (disksing) [inline server/schedule/hot_net.go]: What does _Net_ mean here? How about _HotRegionCache_?

File: server/schedule/hot_net.go [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+	return nil
+}
+
+func (w *HotSpotCache) isRegionHot(id uint64, hotThreshold int) bool {
+	if stat, ok := w.writeFlow.Peek(id); ok {
+		return stat.(*core.RegionStat).HotDegree >= hotThreshold
+	}
+	if stat, ok := w.readFlow.Peek(id); ok {
+		return stat.(*core.RegionStat).HotDegree >= hotThreshold
+	}
```
Comment (disksing) [inline server/schedule/hot_net.go]: What if the id is in both _writeFlow_ and _readFlow_, and in _writeFlow_ it's not hot but in _readFlow_ it's hot?

File: server/cache.go [hunk 3]
```diff
@@ -240,7 +240,15 @@ func (c *clusterInfo) GetRegion(regionID uint64) *core.RegionInfo {
...
-	return c.BasicCluster.IsRegionHot(id)
+	return c.BasicCluster.IsRegionHot(id, c.GetHotRegionLowThreshold())
+}
+
+// RandomHotRegionFromStore random picks a hot region in specify store.
+func (c *clusterInfo) RandomHotRegionFromStore(store uint64, kind schedule.FlowKind) *core.RegionInfo {
+	c.RLock()
+	defer c.RUnlock()
+	r := c.HotNet.RandomHotRegionFromStore(store, kind, c.GetHotRegionLowThreshold())
+	return c.BasicCluster.GetRegion(r.RegionID)
```
Comment (disksing) [inline server/cache.go]: Check nil.

File: server/cache.go [hunk 4]
```diff
@@ -240,7 +240,15 @@ func (c *clusterInfo) GetRegion(regionID uint64) *core.RegionInfo {
...
-	return c.BasicCluster.IsRegionHot(id)
+	return c.BasicCluster.IsRegionHot(id, c.GetHotRegionLowThreshold())
+}
+
+// RandomHotRegionFromStore random picks a hot region in specify store.
+func (c *clusterInfo) RandomHotRegionFromStore(store uint64, kind schedule.FlowKind) *core.RegionInfo {
+	c.RLock()
+	defer c.RUnlock()
+	r := c.HotCache.RandomHotRegionFromStore(store, kind, c.GetHotRegionLowThreshold())
+	return c.BasicCluster.GetRegion(r.RegionID)
```
Comment (disksing) [inline server/cache.go]: Check nil.

File: server/schedulers/mockcluster.go [hunk 5]
```diff
@@ -69,6 +69,17 @@ func (mc *mockCluster) GetStoresAverageScore(kind core.ResourceKind, filters ...
...
 
+// IsRegionHot checks if the region is hot
+func (mc *mockCluster) IsRegionHot(id uint64) bool {
+	return mc.BasicCluster.IsRegionHot(id, mc.GetHotRegionLowThreshold())
+}
+
+// RandomHotRegionFromStore random picks a hot region in specify store.
+func (mc *mockCluster) RandomHotRegionFromStore(store uint64, kind schedule.FlowKind) *core.RegionInfo {
+	r := mc.HotCache.RandomHotRegionFromStore(store, kind, mc.GetHotRegionLowThreshold())
+	return mc.GetRegion(r.RegionID)
```
Comment (disksing) [inline server/schedulers/mockcluster.go]: Check nil.

File: server/schedule/hot_net.go (hunk 2 above)
Comment (nolouch) [reply server/schedule/hot_net.go]: It also returns true now, Do we need to distinguish them?

Comment (disksing) [reply server/schedule/hot_net.go]: I think it will return false.

Comment (nolouch) [reply server/schedule/hot_net.go]: I think we should skip hot region in balance whatever it is hot read or write.

Comment (nolouch) [reply server/schedule/hot_net.go]: got it.

**Comment by disksing** [comment]
LGTM.

File: server/cache.go [hunk 6]
```diff
@@ -240,7 +240,18 @@ func (c *clusterInfo) GetRegion(regionID uint64) *core.RegionInfo {
 func (c *clusterInfo) IsRegionHot(id uint64) bool {
 	c.RLock()
 	defer c.RUnlock()
-	return c.BasicCluster.IsRegionHot(id)
+	return c.BasicCluster.IsRegionHot(id, c.GetHotRegionLowThreshold())
+}
+
+// RandHotRegionFromStore random picks a hot region in specify store.
```
Comment (Connor1996) [inline server/cache.go]: s/random/randomly

Comment (Connor1996) [reply server/cache.go]: s/specify/sepcified

**Comment by nolouch** [comment]
add test cases, PTAL @disksing @Connor1996

**Comment by disksing** [comment]
LGTM.
/run-all-tests

**Pull request merged**